File: packages/core/src/parsers/m3u-parser.ts
import { NOOP_LOGGER, type Logger } from '../logger.js';

export interface M3uEntry {
  duration: number;
  title: string;
  groupTitle: string;
  tvgId: string;
  tvgName: string;
  tvgLogo: string;
  streamUrl: string;
  rawAttributes: string;
  /** Catch-up type: "default", "flussonic", "xc", "shift", "append" */
  catchupType?: string;
  /** URL template for catch-up playback */
  catchupSource?: string;
  /** Catch-up archive window in hours */
  catchupDays?: number;
}

export interface M3uParseResult {
  entries: M3uEntry[];
  /** EPG URL extracted from the #EXTM3U url-tvg header attribute */
  epgUrl?: string;
}

/**
 * Streaming M3U parser. Processes line-by-line to handle large playlists
 * without loading the entire file into memory for parsing.
 *
 * Handles: BOM markers, Windows/Unix line endings, empty lines,
 * malformed entries, and common provider quirks.
 */
export function parseM3u(content: string, logger: Logger = NOOP_LOGGER): M3uParseResult {
  const entries: M3uEntry[] = [];
  // Track URLs we've already added so duplicates from the same playlist
  // don't get classified twice. Titled entries take precedence over bare
  // URL entries (a titled entry inserted later wins by replacing the bare).
  const urlIndex = new Map<string, number>(); // streamUrl -> index in entries
  let duplicates = 0;
  let epgUrl: string | undefined;

  // Strip BOM marker if present
  const cleaned = content.charCodeAt(0) === 0xfeff ? content.slice(1) : content;

  // Normalize line endings and split
  const lines = cleaned.replace(/\r\n/g, '\n').replace(/\r/g, '\n').split('\n');

  let currentEntry: Partial<M3uEntry> | null = null;

  const addOrReplace = (entry: M3uEntry, hasTitle: boolean) => {
    const existingIdx = urlIndex.get(entry.streamUrl);
    if (existingIdx === undefined) {
      urlIndex.set(entry.streamUrl, entries.length);
      entries.push(entry);
      return;
    }
    duplicates++;
    const existing = entries[existingIdx];
    // If the existing entry is a bare URL (no title) and the new one has
    // a real title, upgrade. Otherwise drop the duplicate.
    if (hasTitle && !existing.title) {
      entries[existingIdx] = entry;
    }
  };

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i].trim();

    // Skip empty lines
    if (!line) continue;

    // Parse #EXTM3U header — extract url-tvg for EPG auto-detection
    if (line.startsWith('#EXTM3U')) {
      epgUrl = extractAttribute(line, 'url-tvg') || undefined;
      // Also check x-tvg-url (alternative attribute some providers use)
      if (!epgUrl) {
        epgUrl = extractAttribute(line, 'x-tvg-url') || undefined;
      }
      if (epgUrl) {
        logger.info(`M3U header contains EPG URL: ${epgUrl}`);
      }
      continue;
    }

    // Parse #EXTINF line
    if (line.startsWith('#EXTINF:')) {
      currentEntry = parseExtinfLine(line);
      continue;
    }

    // Skip other directives
    if (line.startsWith('#')) continue;

    // This is a URL line — pair it with the current EXTINF entry
    if (currentEntry) {
      addOrReplace(
        {
          duration: currentEntry.duration ?? -1,
          title: currentEntry.title ?? '',
          groupTitle: currentEntry.groupTitle ?? '',
          tvgId: currentEntry.tvgId ?? '',
          tvgName: currentEntry.tvgName ?? '',
          tvgLogo: currentEntry.tvgLogo ?? '',
          streamUrl: line,
          rawAttributes: currentEntry.rawAttributes ?? '',
          catchupType: currentEntry.catchupType,
          catchupSource: currentEntry.catchupSource,
          catchupDays: currentEntry.catchupDays,
        },
        Boolean(currentEntry.title),
      );
      currentEntry = null;
    } else {
      // URL without preceding EXTINF — create a bare entry
      addOrReplace(
        {
          duration: -1,
          title: extractTitleFromUrl(line),
          groupTitle: '',
          tvgId: '',
          tvgName: '',
          tvgLogo: '',
          streamUrl: line,
          rawAttributes: '',
        },
        false,
      );
    }
  }

  if (duplicates > 0) {
    logger.warn(`Parsed M3U: ${entries.length} unique entries (${duplicates} duplicate URLs collapsed)`);
  } else {
    logger.info(`Parsed ${entries.length} entries from M3U`);
  }
  return { entries, epgUrl };
}

function parseExtinfLine(line: string): Partial<M3uEntry> {
  // Format: #EXTINF:duration tvg-id="..." tvg-name="..." tvg-logo="..." group-title="...",Title
  const entry: Partial<M3uEntry> = {};

  // Extract duration — everything between #EXTINF: and the first space or comma
  const afterPrefix = line.substring(8); // Remove "#EXTINF:"
  const durationMatch = afterPrefix.match(/^(-?\d+(?:\.\d+)?)/);
  entry.duration = durationMatch ? parseFloat(durationMatch[1]) : -1;

  // Extract attributes from the line
  entry.tvgId = extractAttribute(line, 'tvg-id');
  entry.tvgName = extractAttribute(line, 'tvg-name');
  entry.tvgLogo = extractAttribute(line, 'tvg-logo');
  entry.groupTitle = extractAttribute(line, 'group-title');
  entry.rawAttributes = afterPrefix;

  // Extract catch-up attributes (used by some M3U providers)
  const catchupType = extractAttribute(line, 'catchup') || extractAttribute(line, 'catchup-type');
  if (catchupType) entry.catchupType = catchupType;
  const catchupSource = extractAttribute(line, 'catchup-source');
  if (catchupSource) entry.catchupSource = catchupSource;
  const catchupDays = extractAttribute(line, 'catchup-days') || extractAttribute(line, 'tvg-rec');
  if (catchupDays) entry.catchupDays = parseInt(catchupDays, 10) || undefined;

  // Extract title — everything after the last comma
  const lastCommaIndex = line.lastIndexOf(',');
  if (lastCommaIndex !== -1) {
    entry.title = line.substring(lastCommaIndex + 1).trim();
  } else {
    entry.title = '';
  }

  return entry;
}

/**
 * Precompiled regex map for every M3U `#EXTINF` attribute the parser
 * reads. Building the patterns at module load (rather than on every
 * call via `new RegExp(`${attr}=...`)` ) closes Semgrep's
 * detect-non-literal-regexp finding and removes the per-line regex
 * construction cost — `extractAttribute` is called up to 7 times
 * per channel on multi-thousand-channel playlists.
 *
 * Each attribute has two patterns: one matching `attr="..."` (the
 * common form) and one matching `attr='...'`.
 */
const ATTRIBUTE_NAMES = [
  'tvg-id',
  'tvg-name',
  'tvg-logo',
  'group-title',
  'url-tvg',
  'x-tvg-url',
  'catchup',
  'catchup-type',
  'catchup-source',
  'catchup-days',
  'tvg-rec',
] as const;

type AttributeName = (typeof ATTRIBUTE_NAMES)[number];

const ATTRIBUTE_REGEX = new Map<AttributeName, { double: RegExp; single: RegExp }>(
  ATTRIBUTE_NAMES.map((attr) => [
    attr,
    {
      double: new RegExp(`${attr}="([^"]*)"`, 'i'),
      single: new RegExp(`${attr}='([^']*)'`, 'i'),
    },
  ]),
);

function extractAttribute(line: string, attr: AttributeName): string {
  const patterns = ATTRIBUTE_REGEX.get(attr);
  if (!patterns) return '';
  const match = line.match(patterns.double);
  if (match) return match[1];
  const matchSingle = line.match(patterns.single);
  if (matchSingle) return matchSingle[1];
  return '';
}

function extractTitleFromUrl(url: string): string {
  try {
    const pathname = new URL(url).pathname;
    const filename = pathname.split('/').pop() || '';
    // Remove extension
    return filename.replace(/\.[^.]+$/, '').replace(/[_-]/g, ' ');
  } catch {
    return url.split('/').pop() || 'Unknown';
  }
}

File: packages/shared/src/commonMain/kotlin/com/yancotv/shared/parsers/M3uParser.kt
package com.yancotv.shared.parsers

import com.yancotv.shared.logger.Logger
import com.yancotv.shared.logger.NOOP_LOGGER

/**
 * Kotlin port of `@yancotv/core` `m3u-parser.ts`. Behavior parity is required —
 * tests in commonTest mirror `tests/unit/m3u-parser.test.ts`.
 */
data class M3uEntry(
    val duration: Double,
    val title: String,
    val groupTitle: String,
    val tvgId: String,
    val tvgName: String,
    val tvgLogo: String,
    val streamUrl: String,
    val rawAttributes: String,
    val catchupType: String? = null,
    val catchupSource: String? = null,
    val catchupDays: Int? = null,
    /**
     * `catchup-correction="-1.5"` (offset in HOURS). Some providers'
     * recording archives don't line up with the EPG (DST, provider-side
     * TZ misconfig, reseller offset). Player libraries (TiviMate /
     * IPTVnator / Kodi PVR) read this attribute and shift the programme
     * start by `correction * 3600 s` when computing the catch-up URL.
     */
    val catchupCorrection: Double? = null,
)

data class M3uParseResult(
    val entries: List<M3uEntry>,
    /** EPG URL extracted from the #EXTM3U url-tvg / x-tvg-url header attribute. */
    val epgUrl: String? = null,
)

/**
 * Parse a playlist held entirely in memory.
 *
 * **Prefer [parseM3uLines] with a streamed line source for anything
 * provider-sized** — see MB-230. This overload exists for tests, for the
 * TypeScript-parity corpus, and for callers that already hold the text.
 *
 * Handles: BOM markers, Windows/Unix/old-Mac line endings, empty lines,
 * malformed entries, duplicate URL collapse, and common provider quirks.
 */
fun parseM3u(content: String, logger: Logger = NOOP_LOGGER): M3uParseResult = parseM3uLines(content.m3uLineSequence(), logger)

/**
 * Genuinely streaming M3U parser: consumes [lines] lazily and never holds
 * the playlist text.
 *
 * **MB-230 (Critical, 2026-07-25).** The previous implementation took the
 * whole playlist as one `String` and then did
 * `substring` → `replace("\r\n","\n")` → `replace("\r","\n")` → `split('\n')`.
 * Every step allocates a *complete additional copy*, and JVM/ART strings are
 * UTF-16, so a playlist weighing N bytes on the wire peaked at roughly 8-10 N
 * of live heap — all of it reachable at once. On the reported install
 * (255,843 entries) that is several hundred MB against a Fire TV Stick's
 * 384 MB `largeHeap` budget, which is the runaway-GC signature captured in
 * bugs.md: heap pinned at 376/384 MB, 101 consecutive GCs freeing 0 bytes.
 *
 * The line-splitting rules are unchanged and pinned by [M3uParserTest]
 * (`handlesBomMarker`, `handlesCrlfLineEndings`, `handlesCrOnlyLineEndings`):
 * `\r\n`, `\r` and `\n` all terminate a line, and a leading U+FEFF is dropped.
 *
 * What still scales with catalogue size is the returned [M3uParseResult] —
 * one [M3uEntry] per unique stream URL, plus the URL index used to collapse
 * duplicates. That is inherent to the de-duplication contract (a later titled
 * entry replaces an earlier bare-URL one), it is an order of magnitude below
 * the text copies it replaces, and it is what [com.yancotv.shared.sources.BulkContentWriter]
 * consumes in chunks.
 */
fun parseM3uLines(lines: Sequence<String>, logger: Logger = NOOP_LOGGER): M3uParseResult {
    val entries = mutableListOf<M3uEntry>()
    // streamUrl -> index; titled entries replace bare URL entries on collision.
    val urlIndex = mutableMapOf<String, Int>()
    var duplicates = 0
    var epgUrl: String? = null

    var currentEntry: PartialEntry? = null

    fun addOrReplace(entry: M3uEntry, hasTitle: Boolean) {
        val existingIdx = urlIndex[entry.streamUrl]
        if (existingIdx == null) {
            urlIndex[entry.streamUrl] = entries.size
            entries.add(entry)
            return
        }
        duplicates++
        val existing = entries[existingIdx]
        if (hasTitle && existing.title.isEmpty()) {
            entries[existingIdx] = entry
        }
    }

    for (raw in lines) {
        val line = raw.trim()
        if (line.isEmpty()) continue

        if (line.startsWith("#EXTM3U")) {
            epgUrl = extractAttribute(line, "url-tvg").ifEmpty { null }
                ?: extractAttribute(line, "x-tvg-url").ifEmpty { null }
            if (epgUrl != null) logger.info("M3U header contains EPG URL: $epgUrl")
            continue
        }

        if (line.startsWith("#EXTINF:")) {
            currentEntry = parseExtinfLine(line)
            continue
        }

        if (line.startsWith("#")) continue

        val ce = currentEntry
        if (ce != null) {
            addOrReplace(
                M3uEntry(
                    duration = ce.duration ?: -1.0,
                    title = ce.title ?: "",
                    groupTitle = ce.groupTitle ?: "",
                    tvgId = ce.tvgId ?: "",
                    tvgName = ce.tvgName ?: "",
                    tvgLogo = ce.tvgLogo ?: "",
                    streamUrl = line,
                    rawAttributes = ce.rawAttributes ?: "",
                    catchupType = ce.catchupType,
                    catchupSource = ce.catchupSource,
                    catchupDays = ce.catchupDays,
                    catchupCorrection = ce.catchupCorrection,
                ),
                hasTitle = !ce.title.isNullOrEmpty(),
            )
            currentEntry = null
        } else {
            addOrReplace(
                M3uEntry(
                    duration = -1.0,
                    title = extractTitleFromUrl(line),
                    groupTitle = "",
                    tvgId = "",
                    tvgName = "",
                    tvgLogo = "",
                    streamUrl = line,
                    rawAttributes = "",
                ),
                hasTitle = false,
            )
        }
    }

    if (duplicates > 0) {
        logger.warn("Parsed M3U: ${entries.size} unique entries ($duplicates duplicate URLs collapsed)")
    } else {
        logger.info("Parsed ${entries.size} entries from M3U")
    }
    return M3uParseResult(entries = entries.toList(), epgUrl = epgUrl)
}

private class PartialEntry(
    var duration: Double? = null,
    var title: String? = null,
    var groupTitle: String? = null,
    var tvgId: String? = null,
    var tvgName: String? = null,
    var tvgLogo: String? = null,
    var rawAttributes: String? = null,
    var catchupType: String? = null,
    var catchupSource: String? = null,
    var catchupDays: Int? = null,
    var catchupCorrection: Double? = null,
)

private val DURATION_REGEX = Regex("""^(-?\d+(?:\.\d+)?)""")

private fun parseExtinfLine(line: String): PartialEntry {
    val entry = PartialEntry()
    val afterPrefix = line.substring(8) // drop "#EXTINF:"

    val dm = DURATION_REGEX.find(afterPrefix)
    entry.duration = dm?.groupValues?.get(1)?.toDoubleOrNull() ?: -1.0

    entry.tvgId = extractAttribute(line, "tvg-id")
    entry.tvgName = extractAttribute(line, "tvg-name")
    entry.tvgLogo = extractAttribute(line, "tvg-logo")
    entry.groupTitle = extractAttribute(line, "group-title")
    entry.rawAttributes = afterPrefix

    val catchupType = extractAttribute(line, "catchup").ifEmpty { extractAttribute(line, "catchup-type") }
    if (catchupType.isNotEmpty()) entry.catchupType = catchupType
    val catchupSource = extractAttribute(line, "catchup-source")
    if (catchupSource.isNotEmpty()) entry.catchupSource = catchupSource
    val catchupDaysRaw = extractAttribute(line, "catchup-days").ifEmpty { extractAttribute(line, "tvg-rec") }
    if (catchupDaysRaw.isNotEmpty()) entry.catchupDays = catchupDaysRaw.toIntOrNull()
    // `catchup-correction="-1.5"` — offset in HOURS to apply when
    // computing the catch-up URL. Honoured by TiviMate / IPTVnator /
    // Kodi PVR; without it providers whose archives don't line up
    // with the EPG play the wrong slot silently. Audit catch.
    val catchupCorrectionRaw = extractAttribute(line, "catchup-correction")
    if (catchupCorrectionRaw.isNotEmpty()) entry.catchupCorrection = catchupCorrectionRaw.toDoubleOrNull()

    val lastComma = line.lastIndexOf(',')
    entry.title = if (lastComma != -1) line.substring(lastComma + 1).trim() else ""
    return entry
}

private fun extractAttribute(line: String, attr: String): String {
    val escaped = Regex.escape(attr)
    val dbl = Regex("$escaped=\"([^\"]*)\"", RegexOption.IGNORE_CASE).find(line)
    if (dbl != null) return dbl.groupValues[1]
    val sgl = Regex("$escaped='([^']*)'", RegexOption.IGNORE_CASE).find(line)
    if (sgl != null) return sgl.groupValues[1]
    return ""
}

private val URL_SCHEME_REGEX = Regex("""^[a-zA-Z][a-zA-Z0-9+.\-]*://""")

private fun extractTitleFromUrl(url: String): String {
    // Mirrors TS `new URL(url).pathname` fallback. If it has no scheme we
    // fall through to the last-segment path, matching the TS catch branch.
    val hasScheme = URL_SCHEME_REGEX.containsMatchIn(url)
    if (hasScheme) {
        // Find path portion after the authority.
        val afterScheme = url.substringAfter("://", "")
        val pathStart = afterScheme.indexOf('/')
        val pathname = if (pathStart >= 0) afterScheme.substring(pathStart) else ""
        // Strip query/fragment.
        val clean = pathname.substringBefore('?').substringBefore('#')
        val filename = clean.substringAfterLast('/', "")
        return filename.replace(Regex("""\.[^.]+$"""), "").replace(Regex("""[_\-]"""), " ")
    }
    return url.substringAfterLast('/', "").ifEmpty { "Unknown" }
}
